pkg/daemon/persist.go

```diff
@@ -0,0 +1,155 @@
+package daemon
+
+import (
+	"container/heap"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+)
+
+// TaskStorage is a leveldb-backed priority queue.
+// storage is persisted before the in-memory queue to prevent inconsistency between restarts; what
+// you see in the queue is the same as what is written the database
+type TaskStorage struct {
+	Max       int
+	Path      string
+	DBOpts    *opt.Options
+	WriteOpts *opt.WriteOptions
+	ReadOpts  *opt.ReadOptions
+	tq        *TaskQueue
+	db        *leveldb.DB
+	mux       sync.Mutex
+}
+
+// Open database and load its contents into memory.
+func (s *TaskStorage) Open() error {
+	db, err := leveldb.OpenFile(s.Path, s.DBOpts)
+	if err != nil {
+		return err
+	}
+	s.db = db
+	return s.Reload()
+}
+
+// Method for closing the database. Always run Close when finished.
+func (s *TaskStorage) Close() error {
+	s.tq = nil
+	return s.db.Close()
+}
+
+// Read everything from the database into memory. Typically, you will not need to run this.
+func (s *TaskStorage) Reload() error {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	s.tq = new(TaskQueue)
+	iter := s.db.NewIterator(nil, s.ReadOpts)
+	for iter.Next() {
+		tsk := new(Task)
+		err := json.Unmarshal(iter.Value(), tsk)
+		if err != nil {
+			return err
+		}
+		if tsk.State != TaskStateComplete {
+			heap.Push(s.tq, tsk)
+		}
+	}
+	iter.Release()
+	return iter.Error()
+}
+
+// Add an item to the priority queue
+func (s *TaskStorage) Push(tsk *Task) error {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	if s.Len() >= s.Max {
+		return fmt.Errorf("push rejected. too many items.")
+	}
+	tsk.State = TaskStateScheduled
+	key := []byte(tsk.ID)
+	val, err := json.Marshal(tsk)
+	if err != nil {
+		return err
+	}
+	err = s.db.Put(key, val, nil)
+	if err != nil {
+		return err
+	}
+	heap.Push(s.tq, tsk)
+	return nil
+}
+
+// get the next item from the priority queue
+func (s *TaskStorage) Pop() (*Task, error) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	if s.Len() == 0 {
+		return nil, nil
```

Comment:
I think we want to return an error if we try to `Pop` and empty queue. `nil` is a bit dangerous I think.

Reply:
You may be right! I'll return an error here.

Reply:
The idiomatic Go way is to return an `ok bool`, rather than an error. Would that work here?

Reply:
@raulk the way this `Queue` also encapsulates the actual storing, that would mean we have 3 arguments - return value, error, bool

I think it is just fine to return `ErrEmpty` and detect it, rather than add a 3rd argument.

Reply:
Used a named error -- the error would just be used to detect if the queue is popped while empty, so I could be convinced to switch it for a boolean.